File: app/src/main/java/com/bortolan/iquadriv2/Views/ScheduleList.java
package com.bortolan.iquadriv2.Views;


import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bortolan.iquadriv2.Adapters.AdapterOrari;
import com.bortolan.iquadriv2.Databases.FavouritesDB;
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubItem;
import com.bortolan.iquadriv2.R;

import java.util.LinkedList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ScheduleList extends LinearLayout {
    private Context mContext;

    @BindView(R.id.title)
    TextView title;
    @BindView(R.id.recycler)
    RecyclerView recycler;

    private AdapterOrari adapter;
    private List<GitHubItem> items;
    FavouritesDB db;
    AdapterOrari.UpdateFragment updateFragment;

    public ScheduleList(Context context) {
        super(context);
        this.mContext = context;
        init();
    }

    public ScheduleList(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.mContext = context;
        init();
    }

    public ScheduleList(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.mContext = context;
        init();
    }

    public void init() {
        inflate(mContext, R.layout.view_orari, this);
        ButterKnife.bind(this);

        items = new LinkedList<>();

        recycler.setLayoutManager(new LinearLayoutManager(mContext, LinearLayoutManager.HORIZONTAL, false));
        recycler.setHasFixedSize(true);
        recycler.getLayoutManager().setAutoMeasureEnabled(true);
    }

    public void setData(String title, List<GitHubItem> items, FavouritesDB db, AdapterOrari.UpdateFragment updateFragment) {
        this.db = db;
        this.updateFragment = updateFragment;

        if (adapter != null) adapter.clear();
        this.items.clear();
        if (items != null && items.size() != 0) {
            this.title.setText(title.toUpperCase());
            this.items.addAll(items);

            adapter = new AdapterOrari(mContext, db, updateFragment);
            adapter.addAll(items);
            recycler.setAdapter(adapter);
            recycler.addOnScrollListener(new myScrollListener(this.title));

            setVisibility(VISIBLE);
        } else {
            setVisibility(GONE);
        }
    }

    public void filter(String s) {
        if (adapter != null)
            adapter.getFilter().filter(s);
    }


    protected class myScrollListener extends RecyclerView.OnScrollListener {
        private TextView textView;
        String _text;
        int myState, firstVisible;

        public myScrollListener(TextView title) {
            super();
            textView = title;
            _text = title.getText().toString();
        }

        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
            myState = newState;
            if (myState == RecyclerView.SCROLL_STATE_IDLE) textView.setText(_text);
        }

        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);

            if (myState != RecyclerView.SCROLL_STATE_IDLE) {
                firstVisible = ((LinearLayoutManager) recyclerView.getLayoutManager()).findFirstCompletelyVisibleItemPosition();
                //QUANDO FALLISCE IL DOWNLOAD NON CI SONO ITEM -> IndexOutOfBound
                if (((AdapterOrari) recyclerView.getAdapter()).getItem(firstVisible) != null)
                    textView.setText(_text + " - " + ((AdapterOrari) recyclerView.getAdapter()).getItem(firstVisible).getName().substring(0, 2).toUpperCase());
            }

        }
    }
}

File: app/src/main/java/com/bortolan/iquadriv2/Views/ScheduleList.kt
package com.bortolan.iquadriv2.Views

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.bortolan.iquadriv2.Adapters.AdapterOrari
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubItem
import com.bortolan.iquadriv2.R
import java.util.*

open class ScheduleList : LinearLayout {
    internal var updateFragment: AdapterOrari.UpdateFragment? = null
    private lateinit var adapter: AdapterOrari
    private var order: Boolean = false

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    fun init() {
        View.inflate(context, R.layout.view_orari, this)

        recycler.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recycler.setHasFixedSize(true)
        recycler.layoutManager.isAutoMeasureEnabled = true
        adapter = AdapterOrari(context)
        recycler.adapter = adapter
    }

    fun setData(title: String, items: List<GitHubItem>, updateFragment: AdapterOrari.UpdateFragment, order: Boolean) {
        this.updateFragment = updateFragment
        this.order = order
        adapter.setUpdateFragment(updateFragment)

        //set title
        if (!title.isNullOrEmpty()) this.title.text = title.toUpperCase()

        //title callback
        recycler.clearOnScrollListeners()
        recycler.addOnScrollListener(myScrollListener(this.title))

        if (items.isNotEmpty()) {
            if (order) {
                Collections.sort(items) { o1, o2 -> o1.name.compareTo(o2.name, ignoreCase = true) }
            }

            adapter.clear()
            adapter.addAll(items)

            this.visibility = View.VISIBLE
        } else {
            this.visibility = View.GONE
        }
    }

    fun remove(position: Int) {
        adapter.remove(position)
        visibility = if (adapter.totalItemCount() > 0) View.VISIBLE else View.GONE
    }

    fun add(t: GitHubItem, position: Int) {
        adapter.add(t, position)
        visibility = if (adapter.totalItemCount() > 0) View.VISIBLE else View.GONE
    }

    fun filter(s: String?) {
        adapter.filter.filter(s)
        adapter.query = s
    }


    protected inner class myScrollListener internal constructor(private val textView: TextView) : RecyclerView.OnScrollListener() {
        internal var _text: String = textView.text.toString()
        internal var myState: Int = 0
        internal var firstVisible: Int = 0

        override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            myState = newState
            if (myState == RecyclerView.SCROLL_STATE_IDLE) textView.text = _text
        }

        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)

            if (myState != RecyclerView.SCROLL_STATE_IDLE) {
                firstVisible = (recyclerView!!.layoutManager as LinearLayoutManager).findFirstCompletelyVisibleItemPosition()
                if (firstVisible >= 0 && firstVisible < recyclerView.adapter.itemCount)
                    textView.text = String.format("%1s - %2s", _text, (recyclerView.adapter as AdapterOrari).getItem(firstVisible).name.substring(0, 2).toUpperCase())
            }

        }
    }
}
